src/typeEvaluator/typeHelpers.ts

```diff
@@ -110,6 +111,34 @@ export function mapConcrete(
   mapper: (node: ConcreteTypeNode) => TypeNode,
   mergeUnions: (nodes: TypeNode[]) => TypeNode = (nodes) =>
     optimizeUnions({type: 'union', of: nodes}),
+): TypeNode {
+  return mapUnion(
+    node,
+    scope,
+    (node) => {
+      if (node.type === 'unknown') {
+        return node
+      }
+      return mapper(node)
+    },
+    mergeUnions,
+  )
+}
+
+/**
+ * mapUnion extracts a _concrete type_ OR an unknown type from a type node, applies the mapping
+ * function to it and returns. Most notably, this will work through unions
+ * (applying the mapping function for each variant) and inline (resolving the
+ * reference).
+ * This method should _only_ be used if you need to handle unknown types, ie when resolving two sides of an and node, and we don't want to abort if one side is unknown.
+ * In most cases, you should use `mapConcrete` instead.
+ **/
+export function mapUnion(
```

Comment:
A bit annoying to need this, but the reason is that when resolving "and" conditions like `_id == $id && _type == "foo"` we don't want to short circuit the mapper when walking `_id == $id`(resolves to `unknown`).

Alternatively we can use mapUnion in `handleOpCallNode` instead, and there do a unknown check so it at least returns a boolean.
```
function handleOpCallNode(node: OpCallNode, scope: Scope): TypeNode {
  $trace('opcall.node %O', node)
  const lhs = walk({node: node.left, scope})
  const rhs = walk({node: node.right, scope})
  return mapUnion(lhs, scope, (left) =>
    // eslint-disable-next-line complexity, max-statements
    mapUnion(rhs, scope, (right) => {
      $trace('opcall.node.concrete "%s" %O', node.op, {left, right})
      if (left.type === 'unknown' || right.type === 'unknown') {
        return {type: 'boolean'}
      }
```


Then we would resolve the correct type for a projection like:
```
{
"foo": $myParameter == 1
}
```

Which currently resolves to `unknown`

Reply:
And similarly: Comparisons operators are always `boolean | null`. We should _always_ be able to fallback to that.

Reply:
Change this into `function handleOpCallNode` handling unknowns with mapUnion, and `==`/`!=` will now always return a `boolean`, while the other operators short circuits to `unknown` if an `unknown` is given.